### PR TITLE
Implement the metric tracking client from a client

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

For purposes of storage fees and benchmarking we implement a transformer (that is something similar to `ValueSplittingKeyValueStoreClient` and `LruCachingKeyValueClient`) that takes a client and returns another one that does the tracking. The effective implementation in the SDK is deferred to further works.

## Proposal

The code is fairly straightforward. For the `find_keys_by_prefix` I initially went into a wild goose chase
of trying to implement a new `Self::Keys` that updates the sizes as it is processed. This was very complicated
(lifetimes, auxiliary types, RwLockWriteGuard, etc.) and ultimately unsuccessful.

There is much wiggle room in the `MetricStat` type. It could be too detailed or not enough. Also, timings would be inserted or done in another transformer. That is expected to be resolved in the review.

## Test Plan

A number of CI tests were added to test the feature.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-views/src/storage_metric.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+            }
+        }
+        Ok(read)
+    }
+
+    async fn read_multi_key_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, Self::Error> {
+        let mut metric_stat = self.metric_stat.write().await;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs:77]: For the `read_multi_key_bytes` we have a problem with the `keys` that are moved. For the others, we have some problems with the fact that the `Send` properties or lack thereof of the obtained `keys`.

File: linera-views/src/storage_metric.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                    metric_stat.n_deletes += 1;
+                    metric_stat.size_writes += key.len();
+                }
+                WriteOperation::Put { key, value } => {
+                    metric_stat.n_puts += 1;
+                    metric_stat.size_writes += key.len() + value.len();
+                }
+                WriteOperation::DeletePrefix { key_prefix } => {
+                    metric_stat.n_delete_prefix += 1;
+                    metric_stat.size_writes += key_prefix.len();
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs:129]: The problem is that such operations require some global access to the database in order to know how much data is left on the database. That is not easy to do and it certainly has some computational costs associated to it.

File: linera-views/src/storage_metric.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    client: MetricKeyValueClient<MemoryClient>,
+}
+
+impl Default for MetricMemoryClient {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MetricMemoryClient {
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs]: That is possible. I prefer the `MetricMemoryClient` but so be it.

File: linera-views/src/storage_metric.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+        self.client.write_batch(batch, base_key).await
+    }
+
+    async fn clear_journal(&self, base_key: &[u8]) -> Result<(), Self::Error> {
+        self.client.clear_journal(base_key).await
+    }
+}
+
+/// Create a new client
+pub fn get_metric_memory_client() -> MetricKeyValueClient<MemoryClient> {
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:143]: I would make this function and the next one methods of `MetricKeyValueClient`.

Comment (ma2bd) [reply linera-views/src/storage_metric.rs:143]: Why do we special case memory? Is this for testing?

File: linera-views/src/storage_metric.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+};
+use async_lock::RwLock;
+use async_trait::async_trait;
+use std::sync::Arc;
+
+/// Array containing metric data for this client
+/// This can be used for storage fees or for other
+/// benchmarking purposes.
+#[derive(Clone, Copy, Default, Debug, Eq, PartialEq)]
+pub struct MetricStat {
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:17]: `Counters`?

File: linera-views/src/storage_metric.rs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    /// The number of missed cases in read_key and read_multi_key
+    pub n_miss_reads: usize,
+    /// The number of Put in the batches
+    pub n_puts: usize,
+    /// The total number of Delete in the batches
+    pub n_deletes: usize,
+    /// The total number of DeletePrefix in the batches
+    pub n_delete_prefix: usize,
+    /// The total data that went into the Batches
+    pub size_writes: usize,
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:29]: `bytes_written`?

File: linera-views/src/storage_metric.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+    /// The total data being read from
+    pub size_reads: usize,
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:31]: `bytes_read`

File: linera-views/src/storage_metric.rs [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+use crate::{
+    batch::{Batch, WriteOperation},
+    common::{KeyIterable, KeyValueIterable, KeyValueStoreClient},
+    memory::{create_memory_client, MemoryClient},
+};
+use async_lock::RwLock;
+use async_trait::async_trait;
+use std::sync::Arc;
+
+/// Array containing metric data for this client
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:13]: I don't think "metric" is the right word. Probably "metered" (or "measured") is more appropriate.

File: linera-views/src/storage_metric.rs [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+/// The `MetricKeyValueClient` encapsulates a client and creates a new one that
+/// measures the operations being done
+#[derive(Clone)]
+pub struct MetricKeyValueClient<K> {
+    /// The inner client that is called by the metric one
+    pub client: K,
+    /// The data contained in the running of this container
+    pub metric_stat: Arc<RwLock<MetricStat>>,
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:41]: We may want to use [atomic counters](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU32.html#method.fetch_add) and avoid the lock. (And if we have to use a lock, we may not need an async one)

File: linera-views/src/storage_metric.rs [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+            metric_stat.size_reads += key?.len();
+        }
+        Ok(keys)
+    }
+
+    async fn find_key_values_by_prefix(
+        &self,
+        key_prefix: &[u8],
+    ) -> Result<Self::KeyValues, Self::Error> {
+        let mut metric_stat = self.metric_stat.write().await;
```
Comment (ma2bd) [inline linera-views/src/storage_metric.rs:107]: This is taking the lock and keeping it during I/O operations.

**Review by ma2bd — COMMENTED** [review]
Thanks for the PR. I made a few stylish comments but more importantly let's meet with @afck and discuss the use case of storage fees:
* Do we need to keep track of the total storage size? (This implies that we know how much we delete)
* How do we interrupt computations early when the storage "budget" is reached?

File: linera-views/src/storage_metric.rs (hunk 10 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs]: Ok, addressed with atomics as suggested.

File: linera-views/src/storage_metric.rs (hunk 9 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs]: Thank you, and thank you for pointing to the exact relevant entry of `fetch_add`.

File: linera-views/src/storage_metric.rs (hunk 7 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs:56]: Ok.

File: linera-views/src/storage_metric.rs (hunk 6 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs:54]: Ok.

File: linera-views/src/storage_metric.rs (hunk 5 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs]: Ok.

File: linera-views/src/storage_metric.rs (hunk 8 above)
Comment (MathieuDutSik) [reply linera-views/src/storage_metric.rs:38]: Ok.

**Comment by MathieuDutSik** [comment]
It was deemed to be the wrong path and so this PR was cancelled.

**Pull request closed without merging**